rpc/server_test.go

```diff
@@ -130,23 +130,25 @@ func TestServerShortLivedConn(t *testing.T) {
 		deadline = time.Now().Add(10 * time.Second)
 	)
 	for i := 0; i < 20; i++ {
-		conn, err := net.Dial("tcp", listener.Addr().String())
-		if err != nil {
-			t.Fatal("can't dial:", err)
-		}
-		defer conn.Close()
-		conn.SetDeadline(deadline)
-		// Write the request, then half-close the connection so the server stops reading.
-		conn.Write([]byte(request))
-		conn.(*net.TCPConn).CloseWrite()
-		// Now try to get the response.
-		buf := make([]byte, 2000)
-		n, err := conn.Read(buf)
-		if err != nil {
-			t.Fatal("read error:", err)
-		}
-		if !bytes.Equal(buf[:n], []byte(wantResp)) {
-			t.Fatalf("wrong response: %s", buf[:n])
-		}
+		func() {
+			conn, err := net.Dial("tcp", listener.Addr().String())
+			if err != nil {
+				t.Fatal("can't dial:", err)
+			}
+			defer conn.Close()
+			conn.SetDeadline(deadline)
+			// Write the request, then half-close the connection so the server stops reading.
+			conn.Write([]byte(request))
+			conn.(*net.TCPConn).CloseWrite()
+			// Now try to get the response.
+			buf := make([]byte, 2000)
+			n, err := conn.Read(buf)
```

Comment:
Another way to fix this would have been to just move the `defer conn.Close()` to here, like this: 
```
         n, err := conn.Read(buf)
         conn.Close()
         if err != nil{ 
               ...
```
I don't really have an opinion. Your way works too. Any other opinions? cc @fjl

Reply:
Yes, I'd prefer that too (moving the close after read).

I also think the `defer` in loop is not a big deal here. It's a unit test, and the loop has a constant number of iterations.